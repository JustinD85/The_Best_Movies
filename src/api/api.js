// const url = 'http://localhost:3000/api/'
const url = 'https://falsemotive.io/api/'

export const postData = async (options = '', user) => {
  const response = await fetch(`${url}${options}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(user)
  })

  if (response.ok) {
    return response;
  } else {
    throw new Error(`Error posting data: ${response.status}`);
  }
}

export const fetchData = async (options) => {
  const response = await fetch(`${url}${options}`)
  if (response.ok) {
    return await response.json()
  } else {
    throw new Error('Error fetching data')
  }
}