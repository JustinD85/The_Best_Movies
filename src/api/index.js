import apiKey from './movie_api_key'

// const dbUrl = 'http://localhost:3000/api/'
const dbUrl = 'https://falsemotive.io/api/'
const moviesUrl = 'https://api.themoviedb.org/3/movie/'
export const options = (method, data) => ({
  method,
  headers: {
    'Content-Type': 'application/json',
  },
  body: JSON.stringify(data)
})

export const postData = async (path = '', data) => {
  const response = await fetch(`${dbUrl}${path}`, options('POST', data))
  if (!response.ok) throw new Error(`Could not post`)
  const unfilteredData = await response.json()
  return unfilteredData.data
}

export const deleteData = async (path = '', data) => {
  const response = await fetch(`${dbUrl}${path}`, options('DELETE', data))
  if (!response.ok) throw new Error(`Could not delete`)
  const unfilteredData = await response.json()
  return unfilteredData.results
}

export const getData = async (path) => {
  const response = await fetch(`${dbUrl}${path}`)
  if (!response.ok) throw new Error(`${response.statusText}`)
  const unfilteredData = await response.json()
  return unfilteredData.data
}

export const getMovies = async (path) => {
  const response = await fetch(`${moviesUrl}${path}${apiKey}`)
  if (!response.ok) throw new Error(`${response.statusText}`)
  const unfilteredMovies = await response.json()
  return unfilteredMovies.results
}
