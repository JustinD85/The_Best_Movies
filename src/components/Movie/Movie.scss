.Movie {
  margin: 0px 10px;
  color: #ECEBEE;
}

.movie-poster {
  height: 600px;
  border-radius: 15px;
  box-shadow: 0 1px 3px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.24);
  transition: all 0.3s cubic-bezier(.25, .8, .25, 1);

  &:hover {
    height: 602px;
    box-shadow: 0 14px 28px rgba(0, 0, 0, 0.25), 0 10px 10px rgba(0, 0, 0, 0.22);
  }
}

.popup {
  z-index: 1;
  position: absolute;
  left: 4%;
  top: 10%;
  border-radius: 15px;
  background: rgba(0, 0, 0, 0.97);
  color: #ECEBEE;
  padding: 0px 30px 30px 30px;
  height: 64%;
  width: 84%;
}

.movie-title {
  letter-spacing: 2px;
}

.close-popup {
  top: 0%;
  right: 92%;
  color: #ECEBEE;
  font-weight: 700;
  text-decoration: none;
  border: 2px solid #5a5a5a8a;
  padding: 5px 8px;
  border-radius: 100px;
  margin: 10px;
  position: absolute;
}

hr {
  width: 71%;
  padding: 1px;
  margin: 4px auto;
}

.movie-release,
.movie-description,
.movie-vote {
  font-size: 18px;
  width: 74%;
  margin: auto;
}

.favorite-btn {
  display: block;
  margin: auto;
  position: relative;
  top: -91.5%;
  left: 42%;
  font-size: 34px;
  
  &:hover {
    cursor: pointer;
  }
}