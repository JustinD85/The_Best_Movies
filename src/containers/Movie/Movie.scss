.Movie {
  width: 30%;
  border: 1px solid white;
  margin: 10px;
}

.movie-title {
  color: #ECEBEE;
}

.movie-poster {
  height: 300px;
}

.favorite-btn {
  display: block;
  margin: auto;
}