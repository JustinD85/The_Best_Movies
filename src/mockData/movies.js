export const mockMovie = {
    "id": "5a7ee86cc3a368189300b429",
    "iso_639_1": "en",
    "iso_3166_1": "US",
    "key": "2mNMvz85NG8",
    "name": "The Grinch - In Theaters November 9 (TV Spot 1) (HD)",
    "site": "YouTube",
    "size": 720,
    "type": "Teaser",
}

export const mockMoviesResponse = {
    "id": 450465,
    "results": [
        {
            "id": "5a7ee86cc3a368189300b429",
            "iso_639_1": "en",
            "iso_3166_1": "US",
            "key": "2mNMvz85NG8",
            "name": "The Grinch - In Theaters November 9 (TV Spot 1) (HD)",
            "site": "YouTube",
            "size": 720,
            "type": "Teaser",
        }
    ]
}

export const mockMoviesArray = [
    {
        "id": "5a7ee86cc3a368189300b429",
        "iso_639_1": "en",
        "iso_3166_1": "US",
        "key": "2mNMvz85NG8",
        "name": "The Grinch - In Theaters November 9 (TV Spot 1) (HD)",
        "site": "YouTube",
        "size": 720,
        "type": "Teaser",
    }
]