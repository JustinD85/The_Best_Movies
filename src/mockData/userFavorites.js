export const mockUserFavesResponse = {
  "data": [
    {
      "id": 137,
      "movie_id": 428078,
      "user_id": 27,
      "title": "Mortal Engines",
      "poster_path": "/uXJVpPXxZO4L8Rz3IG1Y8XvZJcg.jpg",
      "release_date": "2018-12-05",
      "vote_average": "6",
      "overview": "Set in a world many thousands of years in the future. Earth’s cities now roam the globe on huge wheels, devouring each other in a struggle for ever diminishing resources. On one of these massive Traction Cities, Tom Natsworthy has an unexpected encounter with a mysterious young woman from the Outlands who will change the course of his life forever."
    },
    {
      "id": 139,
      "movie_id": 450465,
      "user_id": 27,
      "title": "Glass",
      "poster_path": "/svIDTNUoajS8dLEo7EosxvyAsgJ.jpg",
      "release_date": "2019-01-16",
      "vote_average": "6.9",
      "overview": "In a series of escalating encounters, security guard David Dunn uses his supernatural abilities to track Kevin Wendell Crumb, a disturbed man who has twenty-four personalities. Meanwhile, the shadowy presence of Elijah Price emerges as an orchestrator who holds secrets critical to both men."
    }
  ],
  "message": "Retrieved All favorites"
}

export const mockUserFavesArray = [
  {
    "id": 137,
    "movie_id": 428078,
    "user_id": 27,
    "title": "Mortal Engines",
    "poster_path": "/uXJVpPXxZO4L8Rz3IG1Y8XvZJcg.jpg",
    "release_date": "2018-12-05",
    "vote_average": "6",
    "overview": "Set in a world many thousands of years in the future. Earth’s cities now roam the globe on huge wheels, devouring each other in a struggle for ever diminishing resources. On one of these massive Traction Cities, Tom Natsworthy has an unexpected encounter with a mysterious young woman from the Outlands who will change the course of his life forever."
  },
  {
    "id": 139,
    "movie_id": 450465,
    "user_id": 27,
    "title": "Glass",
    "poster_path": "/svIDTNUoajS8dLEo7EosxvyAsgJ.jpg",
    "release_date": "2019-01-16",
    "vote_average": "6.9",
    "overview": "In a series of escalating encounters, security guard David Dunn uses his supernatural abilities to track Kevin Wendell Crumb, a disturbed man who has twenty-four personalities. Meanwhile, the shadowy presence of Elijah Price emerges as an orchestrator who holds secrets critical to both men."
  }
]

export const mockUserFavesArrayAfterReducer = [
  {
    "id": 137,
    "movie_id": 428078,
    "user_id": 27,
    "title": "Mortal Engines",
    "poster_path": "/uXJVpPXxZO4L8Rz3IG1Y8XvZJcg.jpg",
    "release_date": "2018-12-05",
    "vote_average": "6",
    "overview": "Set in a world many thousands of years in the future. Earth’s cities now roam the globe on huge wheels, devouring each other in a struggle for ever diminishing resources. On one of these massive Traction Cities, Tom Natsworthy has an unexpected encounter with a mysterious young woman from the Outlands who will change the course of his life forever.",
    "isFavorite": true,
  },
  {
    "id": 139,
    "movie_id": 450465,
    "user_id": 27,
    "isFavorite": true,
    "title": "Glass",
    "poster_path": "/svIDTNUoajS8dLEo7EosxvyAsgJ.jpg",
    "release_date": "2019-01-16",
    "vote_average": "6.9",
    "overview": "In a series of escalating encounters, security guard David Dunn uses his supernatural abilities to track Kevin Wendell Crumb, a disturbed man who has twenty-four personalities. Meanwhile, the shadowy presence of Elijah Price emerges as an orchestrator who holds secrets critical to both men."
  }
]