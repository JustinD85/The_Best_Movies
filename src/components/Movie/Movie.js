import React, { Fragment } from 'react'
import { Link } from 'react-router-dom'
import { postData, getData, deleteData } from '../../api'
import { setShouldPromptLogin, setFavorites, setPopup } from '../../actions'
import { connect } from 'react-redux'
import './Movie.scss'

export const Movie = (
  {
    movie, user, setShouldPromptLogin, setFavorites, isPopup, setPopup
  }
) => {
  const imgUrl = 'https://image.tmdb.org/t/p/w500'
  const checkCanFavorite = async () => {
    await setShouldPromptLogin(!user)
    if (user) {
      const pathToAddFav = '/users/favorites/new'
      const pathToGetFav = `users/${user.id}/favorites`
      const pathToDeleteFav = `${pathToGetFav}/${movie.id}`
      try {
        !movie.isFavorite ?
          await postData(pathToAddFav, movie) :
          await deleteData(pathToDeleteFav, movie)
        const favorites = await getData(pathToGetFav)
        setFavorites(favorites, user.id)
      } catch (error) {
        console.log(error)
      }
    }
  }
  const getImage = () => {
    return <Link to={`/movies/${movie.id}`}  >
      <img
        onClick={() => setPopup(true)}
        src={imgUrl + movie.poster_path}
        className='movie-poster'
        alt='movie poster'></img>
    </Link>
  }
  const getTrailer = () => {
    return <iframe width="560" height="315" title={movie.title}
      src={`https://www.youtube.com/embed/${movie.trailer}`}
      frameBorder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
      allowFullScreen></iframe>
  }
  const _class = isPopup ? 'popup' : 'Movie'

  return (
    <div className={_class} >
      {isPopup && <Fragment>
        <Link className='close-popup' onClick={() => setPopup(false)} to='/'>X</Link>
        <h1 className='movie-title'>{movie.title}</h1>
        <hr/>
        <p className='movie-release'>Released: {movie.release_date}</p>
        <hr/>
        {getTrailer()}
        <hr/>
        <p className='movie-description'>{movie.overview}</p>
        <hr/>
        <p className='movie-vote'>Average rating: {movie.vote_average}</p>
      </Fragment>}

      {!isPopup && getImage()}

      <span className='favorite-btn' onClick={() => checkCanFavorite()}>{movie.isFavorite ? '⭐️' : '☆'}</span>
    </div>
  )
}
const mapStateToProps = (state) => ({
  user: state.user,
})

const mapDispatchToProps = (dispatch) => ({
  setFavorites: (favorites, user_id) => dispatch(setFavorites(favorites, user_id)),
  setShouldPromptLogin: (bool) => dispatch(setShouldPromptLogin(bool)),
  setPopup: (bool) => dispatch(setPopup(bool)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Movie)