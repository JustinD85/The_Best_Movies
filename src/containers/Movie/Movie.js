import React from 'react'
import './Movie.scss'

const Movie = ({movie}) => {
  return (
    <div className='Movie'>
      <h1 className='movie-title'>{movie.title}</h1>
      <p className='movie-title'>{movie.release_date}</p>
    </div>
  )
}

export default Movie;