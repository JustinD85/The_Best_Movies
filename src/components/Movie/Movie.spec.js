import React from 'react'
import { shallow, mount } from 'enzyme'
import { Movie } from './Movie'
import { mockMovie } from '../../mockData/movies'
import { postData, getData, deleteData } from '../../api'
jest.mock('../../api')

describe('Movie', () => {
  let wrapper
  let setShouldPromptLoginMock = jest.fn()
  let setFavoritesMock = jest.fn()
  let setPopupMock = jest.fn()

  beforeEach(() => {

    wrapper = shallow(
      <Movie
        movie={mockMovie}
        user={{ name: 'jon', id: 27 }}
        setShouldPromptLogin={setShouldPromptLoginMock}
        setFavorites={setFavoritesMock}
        setPopup={setPopupMock}
        isPopup={true}
      />
    )

  })

  it('should pass snapshot test', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should call setPopupMock 1 time when link is clicked', () => {
    wrapper.find('Link').simulate('click')
    expect(setPopupMock).toHaveBeenCalledTimes(1)
  })

  it('should call setPopupMock with false when link is clicked', () => {
    wrapper.find('Link').simulate('click')
    expect(setPopupMock).toHaveBeenCalledWith(false)
  })

  it('should call setPopup when a movie is clicked', () => {
    wrapper = shallow(
      <Movie
        movie={mockMovie}
        user={{ name: 'jon', id: 27 }}
        setShouldPromptLogin={setShouldPromptLoginMock}
        setFavorites={setFavoritesMock}
        setPopup={setPopupMock}
        isPopup={false}
      />
    )
    wrapper.find('img').simulate('click')
    expect(setPopupMock).toHaveBeenCalledWith(true)
  })

  it('should call setShouldPromptLoginMock when favorite button is clicked', () => {
    wrapper.find('.favorite-btn').simulate('click')
    expect(setShouldPromptLoginMock).toHaveBeenCalledTimes(1)
  })

  it('should call setFavoritesMock when favorite button is clicked', () => {
    // wrapper.find('.favorite-btn').simulate('click')
    expect(setFavoritesMock).toHaveBeenCalledTimes(1)
  })

  it('should call getData when favorite button is clicked', () => {
    // wrapper.find('.favorite-btn').simulate('click')
    expect(getData).toHaveBeenCalledTimes(1)
  })

  it('should call getData when favorite button is clicked', () => {
    // wrapper.find('.favorite-btn').simulate('click')
    expect(postData).toHaveBeenCalledTimes(1)
  })

  it('should call getData when favorite button is clicked', () => {
    // wrapper.find('.favorite-btn').simulate('click')
    expect(deleteData).toHaveBeenCalledTimes(0)
  })
})
