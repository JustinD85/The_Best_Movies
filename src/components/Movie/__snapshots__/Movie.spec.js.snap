// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie should pass snapshot test 1`] = `
<div
  className="popup"
>
  <Link
    className="close-popup"
    onClick={[Function]}
    replace={false}
    to="/"
  >
    X
  </Link>
  <h1
    className="movie-title"
  />
  <p
    className="movie-release"
  >
    Released: 
  </p>
  <p
    className="movie-description"
  />
  <p
    className="movie-vote"
  >
    Average rating: 
  </p>
  <iframe
    allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"
    allowFullScreen={true}
    frameBorder="0"
    height="315"
    src="https://www.youtube.com/embed/undefined"
    width="560"
  />
  <span
    className="favorite-btn"
    onClick={[Function]}
  >
    â˜†
  </span>
</div>
`;
